### PR TITLE
Handle ellipsis in tuple patterns in match exhaustiveness checking

### PR DESCRIPTION
**Pull request by Veykril**

### CODE REVIEW

File: crates/hir_ty/src/diagnostics/match_check.rs [hunk 1]
```diff
@@ -366,16 +366,17 @@ impl PatStack {
...
-                    // Pat::Wild patterns to `pat_ids`. We should be able to use the
-                    // constructors arity for this, but at the time of writing we aren't
-                    // correctly calculating this arity when ellipsis are present.
-                    return Err(MatchCheckErr::NotImplemented);
+            (Pat::Tuple { args: pat_ids, ellipsis }, &Constructor::Tuple { arity }) => {
+                if let Some(ellipsis) = ellipsis {
+                    let (pre, post) = pat_ids.split_at(ellipsis);
+                    let n_wild_pats = arity.saturating_sub(pat_ids.len());
+                    let pre_iter = pre.iter().map(Into::into);
+                    let wildcards = iter::once(PatIdOrWild::Wild).cycle().take(n_wild_pats);
```
Comment (flodiebold) [inline crates/hir_ty/src/diagnostics/match_check.rs:374]: `iter::repeat`?

File: crates/hir_ty/src/diagnostics/match_check.rs [hunk 2]
```diff
@@ -767,10 +768,12 @@ impl Constructor {
 fn pat_constructor(cx: &MatchCheckCtx, pat: PatIdOrWild) -> MatchCheckResult<Option<Constructor>> {
     let res = match pat.as_pat(cx) {
         Pat::Wild => None,
-        // FIXME somehow create the Tuple constructor with the proper arity. If there are
-        // ellipsis, the arity is not equal to the number of patterns.
-        Pat::Tuple { args: pats, ellipsis } if ellipsis.is_none() => {
-            Some(Constructor::Tuple { arity: pats.len() })
+        Pat::Tuple { args: pats, ellipsis: None } => Some(Constructor::Tuple { arity: pats.len() }),
```
Comment (flodiebold) [inline crates/hir_ty/src/diagnostics/match_check.rs:771]: I think we should probably just always do it like in the ellipsis case? :thinking:

Comment (Veykril) [reply crates/hir_ty/src/diagnostics/match_check.rs:771]: Right there isn't really a point in not doing that 👍

**Pull request merged**